File: cpp/leetcode/ParenthesesScore.cpp
//Leetcode Problem No 856 Score of Parentheses
//Solution written by Xuqiang Fang on 24 June, 2018

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <stack>
#include <queue>

using namespace std;
class Solution{
public:
    int scoreOfParentheses(string S) {
        int p = 0;     
        int t = 0;
        vector<string> ans;
        for(int i=0; i<S.length(); ++i){
            if(S[i] == '('){
                t++;
            }
            else{
                t--;
            }
            if(t == 0){
                ans.push_back(move(S.substr(p, i-p+1)));
                p = i+1;
            }
        }
        return helper(ans);
    }
private:
    int helper(vector<string>& l){
        int ans = 0;
        for(auto& s : l){
            if(s == "()"){
                ans += 1;
            }
            else{
                ans += 2 * scoreOfParentheses(s.substr(1, s.length()-2));
            }
        }
        return ans;
    } 
};

int main(){
    Solution s;
    cout << s.scoreOfParentheses("()") << endl;
    cout << s.scoreOfParentheses("(())") << endl;
    cout << s.scoreOfParentheses("()()") << endl;
    cout << s.scoreOfParentheses("(()(()))") << endl;
    return 0;
}

File: python/leetcode/ParenthesesScore.py
'''
Leetcode problem No 856 Score of Parentheses
Solution written by Xuqiang Fang on 24 June, 2018
'''

class Solution(object):
    def scoreOfParentheses(self, S):
        """
        :type S: str
        :rtype: int
        """
        p = 0; c = 0
        ans = []
        for i in range(len(S)):
            if S[i] == '(':
                c += 1
            else:
                c -= 1
            if c == 0:
                ans.append(S[p:i+1])
                p = i + 1
        return self.helper(ans)
    def helper(self, l):
        ans = 0
        for s in l:
            if s == '()':
                ans += 1
            else:
                ans += 2 * self.scoreOfParentheses(s[1:len(s)-1])
        return ans

def main():
    s = Solution()
    print(s.scoreOfParentheses('()'))
    print(s.scoreOfParentheses('()()'))
    print(s.scoreOfParentheses('(())'))
    print(s.scoreOfParentheses('(()(()))'))
main()
